### PR TITLE
During polling, return error on error instead of continued polling

### PR DESCRIPTION
**Pull request by rramkumar1**
Return error on error instead of continuing polling (and likely running into the same error)

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/k8s-cloud-provider/25)

### CODE REVIEW

**Comment by rramkumar1** [comment]
cc @qfel

File: pkg/cloud/service.go [hunk 1]
```diff
@@ -86,7 +86,7 @@ func (s *Service) pollOperation(ctx context.Context, op operation) error {
 		// returning ctx.Err().
 		select {
 		case <-ctx.Done():
-			klog.V(5).Infof("op.pollOperation(%v, %v) not completed, poll count = %d, ctx.Err = %v", ctx, op, pollCount, ctx.Err())
+      klog.V(5).Infof("op.pollOperation(%v, %v) not completed, poll count = %d, ctx.Err = %v", ctx, op, pollCount, ctx.Err())
```
Comment (bowei) [inline pkg/cloud/service.go]: fix indent?

File: pkg/cloud/service.go [hunk 2]
```diff
@@ -95,16 +95,13 @@ func (s *Service) pollOperation(ctx context.Context, op operation) error {
...
-
-		if done {
-			break
+		switch done, err := op.isDone(ctx); {
+		case err != nil:
+      klog.V(5).Infof("op.isDone(%v) error; op = %v, poll count = %d, err = %v, retrying", ctx, op, pollCount, err)
+			return err
+		case done:
+      klog.V(5).Infof("op.isDone(%v) complete; op = %v, poll count = %d, op.err = %v", ctx, op, pollCount, op.error())
+			return op.error()
```
Comment (bowei) [inline pkg/cloud/service.go]: missing return value when !done?

File: pkg/cloud/service.go [hunk 3]
```diff
@@ -95,16 +95,13 @@ func (s *Service) pollOperation(ctx context.Context, op operation) error {
...
 		klog.V(5).Infof("op.isDone(%v) waiting; op = %v, poll count = %d", ctx, op, pollCount)
 		s.RateLimiter.Accept(ctx, op.rateLimitKey())
-		done, err := op.isDone(ctx)
-		if err != nil {
-			klog.V(5).Infof("op.isDone(%v) error; op = %v, poll count = %d, err = %v, retrying", ctx, op, pollCount, err)
-		}
-
-		if done {
-			break
+		switch done, err := op.isDone(ctx); {
```
Comment (bowei) [inline pkg/cloud/service.go]: switch seems extraneous:
```
done, err := op.isDone(ctx)
if err != nil {
  ...
  return ...
}
if done {
  ...
  return ...
}

```

File: pkg/cloud/service_test.go [hunk 4]
```diff
@@ -22,63 +22,104 @@ import (
...
-	// Verify context is cancelled by now.
-	<-ctxCancelled.Done()
-	// Check that pollOperation returns because the context is cancelled.
-	err = s.pollOperation(ctxCancelled, fo)
-	if err == nil {
-		t.Errorf("pollOperation() = nil, want: %v", ctxCancelled.Err())
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			s := Service{RateLimiter: &NopRateLimiter{}}
+			ctx, cfn := context.WithTimeout(context.Background(), 3*time.Second)
```
Comment (bowei) [inline pkg/cloud/service_test.go]: we should use 1ms or less (I'm assuming test does not rely on actual timeout)

Comment (qfel) [reply pkg/cloud/service_test.go]: The timeout is to have decent failures if the test deadlocks (ie. without the corresponding change in service.go). 1ms looks extremely low, can you really bet there will be enough CPU in testing envs to guarantee everything completes within 1ms? 3s seems much more practical - and it kicks in only if the code deadlocks anyway.

File: pkg/cloud/service.go (hunk 3 above)
Comment (qfel) [reply pkg/cloud/service.go]: Seems like a pretty canonical use of switch over a sequence of ifs, eg. https://golang.org/doc/effective_go.html#switch

File: pkg/cloud/service.go (hunk 2 above)
Comment (qfel) [reply pkg/cloud/service.go]: What do you mean? This function is intended to poll-loop, in particular the previous version continued on !done

File: pkg/cloud/service.go (hunk 1 above)
Comment (rramkumar1) [reply pkg/cloud/service.go]: Done.

File: pkg/cloud/service.go (hunk 2 above)
Comment (rramkumar1) [reply pkg/cloud/service.go]: If !done, then it loops again which is expected right?

Comment (bowei) [reply pkg/cloud/service.go]: ok -- my diff was not in unified so I mismatched the braces

File: pkg/cloud/service.go (hunk 3 above)
Comment (bowei) [reply pkg/cloud/service.go]: sgtm

File: pkg/cloud/service_test.go (hunk 4 above)
Comment (bowei) [reply pkg/cloud/service_test.go]: As long as we aren't having unit tests waiting around for 3 s each, sgtm

**Pull request merged**